### PR TITLE
Add VT sequences input support ##windows

### PR DESCRIPTION
**Pull request by GustavoLCR**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

This makes radare2 compatible with the new Windows Terminal and semi-working under MinTTY

* Fix radare2 immediately exiting when console is redirected to pipe
* Hack to get viewport size under MinTTY
* Fix mouse when VT input disabled and buffer scrolled
* Fix line wrapping when VT output disabled

**Test plan**

- On ConHost (Windows 10 1909)
- [x] Test mouse/arrows with `e scr.vtmode=0`
- [x] Test mouse/arrows with `e scr.vtmode=1`
- [x] Test mouse/arrows with `e scr.vtmode=2`

- On Windows Terminal/Mintty
- [x] Test mouse/arrows with `e scr.vtmode=2`

**Closing issues**

Maybe #15674 cc @pandasauce

### CODE REVIEW

File: libr/cons/cons.c [hunk 1]
```diff
@@ -1279,17 +1285,61 @@ R_API bool r_cons_isatty() {
...
 
+#if __WINDOWS__
+static int __xterm_get_cur_pos(int *xpos) {
+	int ypos = 0;
+	(void)write (I.fdout, R_CONS_GET_CURSOR_POSITION, sizeof (R_CONS_GET_CURSOR_POSITION));
+	int ch = r_cons_readchar ();
+	if (ch == 0x1b) {
+		(void)r_cons_readchar ();
+		char pos[16] = { 0 };
+		int i;
```
Comment (radare) [inline libr/cons/cons.c:1296]: use size_t

File: libr/cons/cons.c [hunk 2]
```diff
@@ -1279,17 +1285,61 @@ R_API bool r_cons_isatty() {
 	return false;
 }
 
+#if __WINDOWS__
+static int __xterm_get_cur_pos(int *xpos) {
+	int ypos = 0;
+	(void)write (I.fdout, R_CONS_GET_CURSOR_POSITION, sizeof (R_CONS_GET_CURSOR_POSITION));
```
Comment (radare) [inline libr/cons/cons.c:1291]: check for return value and return early in case the terminal is not there

File: libr/cons/cons.c [hunk 3]
```diff
@@ -1446,9 +1511,17 @@ R_API void r_cons_set_raw(bool is_raw) {
 	}
 #elif __WINDOWS__
 	if (is_raw) {
-		SetConsoleMode (h, I.term_raw);
+		if (I.term_xterm) {
+			system ("stty raw -echo");
```
Comment (radare) [inline libr/cons/cons.c:1515]: use r_sandbox_system() instead

File: libr/cons/cons.c [hunk 4]
```diff
@@ -1446,9 +1511,17 @@ R_API void r_cons_set_raw(bool is_raw) {
...
+		} else {
+			SetConsoleMode (h, I.term_raw);
+		}
 	} else {
-		SetConsoleMode (h, I.term_buf);
+		if (I.term_xterm) {
+			system ("stty -raw -echo");
```
Comment (radare) [inline libr/cons/cons.c:1521]: same

**Comment by pandasauce** [comment]
Does not affect either of the symptoms in #15674 :(

With regards to that ticket, the only noticeable change is with `e scr.vtmode=0` on mintty:

image

**Comment by kazarmy** [comment]
Does setting `e scr.ansicon=1` and running `r2` in `mintty` directly (without `winpty`) help with #15674?

**Comment by pandasauce** [comment]
> Does setting `e scr.ansicon=1` and running `r2` in `mintty` directly (without `winpty`) help with #15674?

There is no `e scr.ansicon`. Setting `e scr.color=1` helps when running it in `winpty`, but obviously means no truecolor.

Running it without `winpty` results in `r2` immediately exiting in all cases.

**Comment by GustavoLCR** [comment]
@pandasauce It works for me under mintty (git bash):

image

**Comment by kazarmy** [comment]
@pandasauce Works for me too, though I lose echo at the bash prompt after quitting. Is your mintty set to emulate xterm? Mintty version?

**Comment by kazarmy** [comment]
I also lose the mouse wheel lol

**Comment by pandasauce** [comment]
Nailed it down after glancing through code.

My `$TERM` is almost always overridden to `screen-256color` by `tmux` (for tabs and splits on `mintty`), which is why it was exiting.

With `$TERM` forced to `xterm` or `xterm-256color`, truecolor works correctly and it runs without `winpty`. If using bash, local echo dies. With zsh it doesn't die, but mouse clicks start appearing in console like this `0;109;20M0;109;20m`.

The `xterm` caveat sounds like something worth documenting, maybe in https://github.com/radareorg/radare2/blob/ff71c41/doc/windows.md ?

This closes #15674, but introduces the local echo/click printing issue.

**Comment by radare** [comment]
Ready to merge?

**Comment by kazarmy** [comment]
Should be mergeable ... I have some fixes lined up.

**Pull request merged**